tempodb/tempodb.go

```diff
@@ -202,66 +202,32 @@ func (rw *readerWriter) WriteBlock(ctx context.Context, c WriteableBlock) error
 }
 
 // CompleteBlock iterates the given WAL block and flushes it to the TempoDB backend.
-func (rw *readerWriter) CompleteBlock(block *wal.AppendBlock, combiner model.ObjectCombiner) (*v2.BackendBlock, error) {
+func (rw *readerWriter) CompleteBlock(block *wal.AppendBlock, combiner model.ObjectCombiner) (common.BackendBlock, error) {
 	return rw.CompleteBlockWithBackend(context.TODO(), block, combiner, rw.r, rw.w)
 }
 
 // CompleteBlock iterates the given WAL block but flushes it to the given backend instead of the default TempoDB backend. The
 // new block will have the same ID as the input block.
-func (rw *readerWriter) CompleteBlockWithBackend(ctx context.Context, block *wal.AppendBlock, combiner model.ObjectCombiner, r backend.Reader, w backend.Writer) (*v2.BackendBlock, error) {
-	meta := block.Meta()
-	blockID := meta.BlockID
-	tenantID := meta.TenantID
-
-	// Default and nil check is primarily to make testing easier.
-	flushSize := DefaultFlushSizeBytes
-	if rw.compactorCfg != nil && rw.compactorCfg.FlushSizeBytes > 0 {
-		flushSize = rw.compactorCfg.FlushSizeBytes
-	}
-
-	iter, err := block.Iterator(combiner)
-	if err != nil {
-		return nil, errors.Wrap(err, "error getting completing block iterator")
-	}
-	defer iter.Close()
+func (rw *readerWriter) CompleteBlockWithBackend(ctx context.Context, block *wal.AppendBlock, combiner model.ObjectCombiner, r backend.Reader, w backend.Writer) (common.BackendBlock, error) {
 
-	newBlock, err := v2.NewStreamingBlock(rw.cfg.Block, blockID, tenantID, []*backend.BlockMeta{meta}, meta.TotalObjects)
+	iter, err := block.TraceIterator(combiner)
 	if err != nil {
-		return nil, errors.Wrap(err, "error creating compactor block")
+		return nil, err
 	}
 
-	var tracker backend.AppendTracker
-	for {
-		id, data, err := iter.Next(ctx)
-		if err != nil && err != io.EOF {
-			return nil, errors.Wrap(err, "error iterating")
-		}
-
-		if id == nil {
-			break
-		}
-
-		err = newBlock.AddObject(id, data)
-		if err != nil {
-			return nil, errors.Wrap(err, "error adding object to compactor block")
-		}
+	meta := block.Meta()
 
-		if newBlock.CurrentBufferLength() > int(flushSize) {
-			tracker, _, err = newBlock.FlushBuffer(ctx, tracker, w)
-			if err != nil {
-				return nil, errors.Wrap(err, "error flushing compactor block")
-			}
-		}
-	}
+	// Pinning to v2 here
```

Comment:
Good catch, this function had several mistakes.  Updated to use latest as suggested, but also fixed to use the latest dataEncoding and honor the backend block encoding config instead of reusing the wal's encoding.